nucliadb/nucliadb/ingest/maindb/pg.py

```diff
@@ -0,0 +1,187 @@
+        async for record in self.connection.cursor(
+            "SELECT key FROM resources WHERE key LIKE $1 LIMIT $2", prefix + "%", limit
+        ):
+            if not include_start and record["key"] == prefix:
+                continue
+            yield record["key"]
+
+
+class PGTransaction(Transaction):
+    driver: PGDriver
+
+    def __init__(self, connection: asyncpg.Connection, txn: Any, driver: PGDriver):
+        self.connection = connection
+        self.data_layer = DataLayer(connection)
+        self.txn = txn
+        self.driver = driver
+        self.open = True
+        self._lock = asyncio.Lock()
+
+    async def abort(self):
+        async with self._lock:
+            if self.open:
+                try:
+                    await self.txn.rollback()
+                finally:
+                    self.open = False
+                    await self.connection.close()
+
+    async def commit(
+        self,
+        worker: Optional[str] = None,
+        tid: Optional[int] = None,
+        resource: bool = True,
+    ):
+        async with self._lock:
+            try:
+                if resource:
+                    if worker is None or tid is None:
+                        raise NoWorkerCommit()
+                    if tid != -1:
+                        # If tid == -1 means we are injecting a resource via gRPC. We don't
+                        # want to save the tid in this case, as it would break the next
+                        # transactionability check.
+                        key = TXNID.format(worker=worker)
+                        await self.data_layer.set(key, str(tid).encode())
+                await self.txn.commit()
+            except Exception:
+                await self.txn.rollback()
+                raise
+            finally:
+                self.open = False
+                await self.connection.close()
+
+    async def batch_get(self, keys: List[str]):
+        return await self.data_layer.batch_get(keys)
+
+    async def get(self, key: str) -> Optional[bytes]:
+        return await self.data_layer.get(key)
+
+    async def set(self, key: str, value: bytes):
+        await self.data_layer.set(key, value)
+
+    async def delete(self, key: str):
+        await self.data_layer.delete(key)
+
+    def keys(
+        self,
+        match: str,
+        count: int = DEFAULT_SCAN_LIMIT,
+        include_start: bool = True,
+    ):
+        return self.data_layer.scan_keys(match, count, include_start=include_start)
+
+
+class PGDriver(Driver):
+    pool: asyncpg.Pool
+
+    def __init__(self, url: str):
+        self.url = url
+        self._lock = asyncio.Lock()
+
+    async def initialize(self):
+        async with self._lock:
+            if self.initialized is False:
+                self.pool = await asyncpg.create_pool(self.url)
+
+            # check if table exists
+            async with self.pool.acquire() as conn:
+                await conn.execute(CREATE_TABLE)
```

Comment:
NIT: should these two lines be nested under `if self.initialized is False` condition?